### ISSUE TITLE
Input data axis ordering

### ISSUE THREAD

**Issue by galalalala** [MIC-DKFZ/medicaldetectiontoolkit#19]
Hi,
 I have a question regarding the shape of input data. In `lidc_exp/data_loader.py`, the class <`BatchGenerator`> will return an dictionary with order (b, c, x, y, (z)). But in `models/backbone.py`, the `forward` fucntion will take images of shape (b, c, y, x, (z)). Where is the conversion between the shapes done and is there a reason for it? And is there a reason for not using the standard order (b, c, (z), y, x)? Thanks.

**Comment by pfjaeger** [comment]
Hi, thanks a lot for pointing this out! In fact there is NO conversion. The order is always (y, x (z)). So the doc string in the lidc data loader is wrong (I will fix this in the next commit)

As for your question on why it is not zyx. There is a good reason:  this framework handles 2D and 3D inputs dynamically. This is much easier to implement if we can append or check for the z dimension at the end of an array/list instead of squeezing it in the middle.

**Comment by galalalala** [comment]
Thanks for replying. I have never tried training detection model in (y,x,z) order before. It will be interesting to try it out. In addition, how is the efficiency of the multithread dataloader in the code? I did some experiments(batch size = 8 and default parameters in lidc config file)  and some batches are lightening fast(~0.1s) but some are quite slow(~20s). Do you have any idea what causes this huge discrepancy?  Thanks a lot.

**Comment by pfjaeger** [comment]
"I have never tried training detection model in (y,x,z) order before. It will be interesting to try it out." I am not quite sure what you mean. The actual order of the 3 dimensions does not matter for training, it is just an implementation decision. Or do you mean it is your first time training Object Detectors in 3 dimensions at all? This should be new, as far as I know, there aren't other implementations of 3D Retina Net or 3D Mask R-CNN.

The huge time discrepancies between batches are not normal. In my setup training runs smoothly with equally timed batches. It sounds like you are in a hardware bottleneck, such as reading from a hard drive or available cpu cores. 

I will close this issue. Feel free to join our new slack channel for further discussion:
https://join.slack.com/t/mdtoolkit/shared_invite/enQtNTQ3MjY2MzE0MDg2LWNjY2I2Njc5MTY0NmM0ZWIxNmQwZDRhYzk2MDdhM2QxYjliYTcwYzhkNTAxYmRkMDA0MjcyNDMyYjllNTZhY2M
